Add an API skeleton to request a <model> element immersive

jeanhbr opened the pull request:
#### 565d1f9077ab94370bd49e47a9c62144b1eb9bc1
<pre>
Add an API skeleton to request a &lt;model&gt; element immersive
<a href="https://bugs.webkit.org/show_bug.cgi?id=302400">https://bugs.webkit.org/show_bug.cgi?id=302400</a>
<a href="https://rdar.apple.com/164630049">rdar://164630049</a>

Reviewed by Etienne Segonzac and Mike Wyrzykowski.

Expose the interface to request a model element as immersive.
As proposed in the following model element issue:
<a href="https://github.com/immersive-web/model-element/issues/133">https://github.com/immersive-web/model-element/issues/133</a>
Original explainer for spatial backdrop:
<a href="https://github.com/WebKit/explainers/tree/main/spatial-backdrop">https://github.com/WebKit/explainers/tree/main/spatial-backdrop</a>

Add a Javascript API on the model element and document.
Add a CSS pseudo selector.

Test: model-element/model-element-immersive.html

* LayoutTests/model-element/model-element-immersive-expected.txt: Added.
* LayoutTests/model-element/model-element-immersive.html: Added.
Add first basic tests for the immersive API.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
Add a feature flag.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add files that will host the immersive logic.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::immersive const):
(WebCore::HTMLModelElement::requestImmersive):
(WebCore::HTMLModelElement::removedFromAncestor):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
Add the Javascript API interface on &lt;model&gt;

* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesImmersivePseudoClass):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
Add the CSS Pseudo selector plumbing for immersive.

* Source/WebCore/dom/Document+Immersive.idl: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::m_syncData):
(WebCore::Document::removedLastRef):
(WebCore::Document::commonTeardown):
(WebCore::Document::ensureImmersive):
(WebCore::Document::immersive):
(WebCore::Document::immersive const):
(WebCore::Document::protectedImmersive):
(WebCore::Document::protectedImmersive const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::immersiveIfExists):
(WebCore::Document::immersiveIfExists const):
* Source/WebCore/dom/DocumentImmersive.cpp: Added.
(WebCore::DocumentImmersive::DocumentImmersive):
(WebCore::DocumentImmersive::immersiveEnabled):
(WebCore::DocumentImmersive::immersiveElement):
(WebCore::DocumentImmersive::immersiveElement const):
(WebCore::DocumentImmersive::exitImmersive):
(WebCore::DocumentImmersive::requestImmersive):
(WebCore::DocumentImmersive::exitRemovedImmersiveElement):
(WebCore::DocumentImmersive::dispatchPendingEvents):
(WebCore::DocumentImmersive::queueImmersiveEventForElement):
(WebCore::DocumentImmersive::clear):
* Source/WebCore/dom/DocumentImmersive.h: Added.
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/PseudoElement.cpp: Unified build fix
* Source/WebCore/dom/ScriptElement.cpp: Unified build fix
(WebCore::reportSpeculationRulesError):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/Logging.h:

Canonical link: <a href="https://commits.webkit.org/303504@main">https://commits.webkit.org/303504@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d4d2f671979c5b43aa46b544f4c4d5559941c23

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132599 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5094 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43663 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140120 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84614 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134469 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5290 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4853 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101372 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68666 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135545 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118782 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82170 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3613 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1374 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83355 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124669 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112539 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36898 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142775 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131107 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4764 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37486 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109749 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4846 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4120 "Found 1 new test failure: imported/w3c/web-platform-tests/mediacapture-streams/overconstrained_error.https.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109930 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27869 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3626 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115054 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/58204 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4818 "Built successfully") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164074 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4654 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68269 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42617 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4909 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4775 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -635,6 +635,10 @@
 #endif
 
+#if !defined(ENABLE_MODEL_ELEMENT_IMMERSIVE) && PLATFORM(VISION)
+#define ENABLE_MODEL_ELEMENT_IMMERSIVE ENABLE_MODEL_PROCESS
```
Do we want to enable this right away? `ENABLE_MODEL_PROCESS` is 1 on shipping visionOS

mwyrzykowski replied:
Oh nevermind, there is a runtime flag too which defaults off. 

This looks great.

mwyrzykowski reviewed (commented):
Overall I think this looks good, I added a few other WebKit reviewers to ensure there are no objections.

The web spec work is on-going, but this is all runtime disabled so I think that is fine.

If there are no objections, I will perform a closer review and likely approve.

mwyrzykowski reviewed (approved):
Since this is a prototype and largely behind a runtime disabled switch, I see no reason to hold up merging the PR while we iterate on this.

etiennesegonzac reviewed (approved):
Thanks for the extra standards discussion links in the commit message. Waiting on EWS but this is ready to go!

jeanhbr committed:
Add an API skeleton to request a <model> element immersive
https://bugs.webkit.org/show_bug.cgi?id=302400
rdar://164630049

Reviewed by Etienne Segonzac and Mike Wyrzykowski.

Expose the interface to request a model element as immersive.
As proposed in the following model element issue:
https://github.com/immersive-web/model-element/issues/133
Original explainer for spatial backdrop:
https://github.com/WebKit/explainers/tree/main/spatial-backdrop

Add a Javascript API on the model element and document.
Add a CSS pseudo selector.

Test: model-element/model-element-immersive.html

* LayoutTests/model-element/model-element-immersive-expected.txt: Added.
* LayoutTests/model-element/model-element-immersive.html: Added.
Add first basic tests for the immersive API.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
Add a feature flag.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add files that will host the immersive logic.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::immersive const):
(WebCore::HTMLModelElement::requestImmersive):
(WebCore::HTMLModelElement::removedFromAncestor):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
Add the Javascript API interface on <model>

* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesImmersivePseudoClass):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
Add the CSS Pseudo selector plumbing for immersive.

* Source/WebCore/dom/Document+Immersive.idl: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::m_syncData):
(WebCore::Document::removedLastRef):
(WebCore::Document::commonTeardown):
(WebCore::Document::ensureImmersive):
(WebCore::Document::immersive):
(WebCore::Document::immersive const):
(WebCore::Document::protectedImmersive):
(WebCore::Document::protectedImmersive const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::immersiveIfExists):
(WebCore::Document::immersiveIfExists const):
* Source/WebCore/dom/DocumentImmersive.cpp: Added.
(WebCore::DocumentImmersive::DocumentImmersive):
(WebCore::DocumentImmersive::immersiveEnabled):
(WebCore::DocumentImmersive::immersiveElement):
(WebCore::DocumentImmersive::immersiveElement const):
(WebCore::DocumentImmersive::exitImmersive):
(WebCore::DocumentImmersive::requestImmersive):
(WebCore::DocumentImmersive::exitRemovedImmersiveElement):
(WebCore::DocumentImmersive::dispatchPendingEvents):
(WebCore::DocumentImmersive::queueImmersiveEventForElement):
(WebCore::DocumentImmersive::clear):
* Source/WebCore/dom/DocumentImmersive.h: Added.
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/PseudoElement.cpp: Unified build fix
* Source/WebCore/dom/ScriptElement.cpp: Unified build fix
(WebCore::reportSpeculationRulesError):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::operator<<):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/Logging.h:

Canonical link: https://commits.webkit.org/303504@main
